diff -ru evolution-3.20.5-orig/e-util/e-plugin.c evolution-3.20.5/e-util/e-plugin.c
--- evolution-3.20.5-orig/e-util/e-plugin.c	2015-03-23 08:12:18.000000000 +0100
+++ evolution-3.20.5/e-util/e-plugin.c	2016-12-05 19:44:15.521479078 +0100
@@ -65,6 +65,8 @@
 
 /* global table of plugin types by pluginclass.type */
 static GHashTable *ep_types;
+/* plugin load path */
+static GSList *ep_path;
 /* global table of plugins by plugin.id */
 static GHashTable *ep_plugins;
 /* the list of disabled plugins from GSettings */
@@ -286,6 +288,7 @@
 e_plugin_class_init (EPluginClass *class)
 {
 	GObjectClass *object_class;
+	gchar *path, *col, *p;
 
 	object_class = G_OBJECT_CLASS (class);
 	object_class->set_property = ep_set_property;
@@ -304,6 +307,22 @@
 			"Whether the plugin is enabled",
 			TRUE,
 			G_PARAM_READWRITE));
+
+	/* Add paths in the environment variable or default global
+	 * and user specific paths */
+	path = g_strdup(g_getenv("EVOLUTION_PLUGIN_PATH"));
+	if (path == NULL) {
+		path = g_strdup(EVOLUTION_PLUGINDIR);
+	}
+
+	p = path;
+	while ((col = strchr (p, G_SEARCHPATH_SEPARATOR))) {
+		*col++ = 0;
+		e_plugin_add_load_path (p);
+		p = col;
+	}
+	e_plugin_add_load_path (p);
+	g_free (path);
 }
 
 static void
@@ -427,6 +446,26 @@
 	return 0;
 }
 
+/**
+ * e_plugin_add_load_path:
+ * @path: The path to add to search for plugins.
+ *
+ * Add a path to be searched when e_plugin_load_plugins() is called.
+ * By default the system plugin directory and ~/.eplugins is used as
+ * the search path unless overriden by the environmental variable
+ * %EVOLUTION_PLUGIN_PATH.
+ *
+ * %EVOLUTION_PLUGIN_PATH is a : separated list of paths to search for
+ * plugin definitions in order.
+ *
+ * Plugin definitions are XML files ending in the extension ".eplug".
+ **/
+void
+e_plugin_add_load_path (const gchar *path)
+{
+	ep_path = g_slist_append (ep_path, g_strdup (path));
+}
+
 static void
 plugin_load_subclass (GType type,
                       GHashTable *hash_table)
@@ -486,6 +525,7 @@
 e_plugin_load_plugins (void)
 {
 	GSettings *settings;
+	GSList *l;
 	gchar **strv;
 	gint i;
 
@@ -514,9 +554,11 @@
 	g_object_unref (settings);
 
 	for (i = 0; i < 3; i++) {
+		for (l = ep_path; l; l = g_slist_next (l)) {
+
 		GDir *dir;
 		const gchar *d;
-		const gchar *path = EVOLUTION_PLUGINDIR;
+		const gchar *path = l->data;
 
 		pd (printf ("scanning plugin dir '%s'\n", path));
 
@@ -537,6 +579,7 @@
 		}
 
 		g_dir_close (dir);
+		}
 	}
 
 	return 0;
diff -ru evolution-3.20.5-orig/e-util/e-plugin.h evolution-3.20.5/e-util/e-plugin.h
--- evolution-3.20.5-orig/e-util/e-plugin.h	2014-03-24 10:25:23.000000000 +0100
+++ evolution-3.20.5/e-util/e-plugin.h	2016-12-05 19:37:35.615265280 +0100
@@ -140,6 +140,7 @@
 GType		e_plugin_get_type		(void) G_GNUC_CONST;
 gint		e_plugin_construct		(EPlugin *plugin,
 						 xmlNodePtr root);
+void		e_plugin_add_load_path		(const gchar *path);
 gint		e_plugin_load_plugins		(void);
 GSList *	e_plugin_list_plugins		(void);
 gpointer	e_plugin_get_symbol		(EPlugin *plugin,
